packages/tokens/fungible/src/extensions/allowlist/storage.rs

```diff
@@ -0,0 +1,212 @@
+use soroban_sdk::{contracttype, panic_with_error, Address, Env};
+use stellar_constants::{BALANCE_EXTEND_AMOUNT, BALANCE_TTL_THRESHOLD};
+
+use crate::{
+    extensions::allowlist::{emit_user_allowed, emit_user_disallowed},
+    overrides::{Base, ContractOverrides},
+    FungibleTokenError,
+};
+
+pub struct AllowList;
+
+impl ContractOverrides for AllowList {
+    fn transfer(e: &Env, from: &Address, to: &Address, amount: i128) {
+        AllowList::transfer(e, from, to, amount);
+    }
+
+    fn transfer_from(e: &Env, spender: &Address, from: &Address, to: &Address, amount: i128) {
+        AllowList::transfer_from(e, spender, from, to, amount);
+    }
+
+    fn approve(e: &Env, owner: &Address, spender: &Address, amount: i128, live_until_ledger: u32) {
+        AllowList::approve(e, owner, spender, amount, live_until_ledger);
+    }
+}
+
+/// Storage keys for the data associated with the allowlist extension
+#[contracttype]
+pub enum AllowListStorageKey {
+    /// Stores the allowed status of an account
+    Allowed(Address),
+}
+
+impl AllowList {
+    // ################## QUERY STATE ##################
+
+    /// Returns the allowed status of an account.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - Access to the Soroban environment.
+    /// * `account` - The address to check the allowed status for.
+    pub fn allowed(e: &Env, account: &Address) -> bool {
+        let key = AllowListStorageKey::Allowed(account.clone());
+        e.storage().persistent().get(&key).unwrap_or(false)
+    }
+
+    // ################## CHANGE STATE ##################
+
+    /// Allows a user to receive and transfer tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `e` - Access to the Soroban environment.
+    /// * `user` - The address to allow.
+    ///
+    /// # Events
+    ///
+    /// * topics - `["user_allowed", user: Address]`
+    /// * data - `[]`
+    ///
+    /// # Security Warning
+    ///
+    /// **IMPORTANT**: This function bypasses authorization checks and should
+    /// only be used:
+    /// - During contract initialization/construction
+    /// - In admin functions that implement their own authorization logic
+    ///
+    /// Using this function in public-facing methods creates significant
+    /// security risks as it could allow unauthorized allowlist
+    /// modifications.
+    pub fn allow_user(e: &Env, user: &Address) {
+        // Set the user as allowed
+        let key = AllowListStorageKey::Allowed(user.clone());
+        e.storage().persistent().set(&key, &true);
+        e.storage().persistent().extend_ttl(&key, BALANCE_TTL_THRESHOLD, BALANCE_EXTEND_AMOUNT);
```

Comment:
we don't need to extend persistent storage when it gets set, only when it's read